File: apps/azimuth/src/routes/atlas/versions.rs
use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use database::models::atlas as db;
use serde::Serialize;
use types::atlas as api;
use util::{ApiError, ApiResult};
use utoipa::ToSchema;

use super::{find_project, find_version, version_response};
use crate::auth::secret::MachineAuth;
use crate::state::AppState;

#[derive(Debug, Serialize, ToSchema)]
pub struct Created {
    pub message: String,
    pub version: String,
}

#[utoipa::path(
    get,
    path = "/atlas/projects/{project}/versions",
    tag = "atlas",
    params(("project" = String, Path, description = "Project key")),
    responses((status = 200, body = Vec<api::VersionResponse>), (status = 404))
)]
pub async fn list(
    State(state): State<AppState>,
    Path(project): Path<String>,
) -> ApiResult<Json<Vec<api::VersionResponse>>> {
    let project = find_project(&state, &project).await?;
    let versions = db::versions_of(&state.db, &[project.id]).await?;

    let mut responses = Vec::with_capacity(versions.len());
    for version in versions {
        let builds = db::build_numbers(&state.db, version.id).await?;
        responses.push(version_response(version, builds));
    }

    Ok(Json(responses))
}

#[utoipa::path(
    get,
    path = "/atlas/projects/{project}/versions/{version}",
    tag = "atlas",
    params(
        ("project" = String, Path, description = "Project key"),
        ("version" = String, Path, description = "Version key")
    ),
    responses((status = 200, body = api::VersionResponse), (status = 404))
)]
pub async fn get(
    State(state): State<AppState>,
    Path((project, version)): Path<(String, String)>,
) -> ApiResult<Json<api::VersionResponse>> {
    let project = find_project(&state, &project).await?;
    let version = find_version(&state, &project, &version).await?;
    let builds = db::build_numbers(&state.db, version.id).await?;

    Ok(Json(version_response(version, builds)))
}

#[utoipa::path(
    post,
    path = "/atlas/projects/{project}/versions/create",
    tag = "atlas",
    security(("api_secret" = [])),
    params(("project" = String, Path, description = "Project key")),
    request_body = api::CreateVersionBody,
    responses((status = 201, body = Created), (status = 404), (status = 409))
)]
pub async fn create(
    State(state): State<AppState>,
    Path(project): Path<String>,
    _auth: MachineAuth,
    Json(body): Json<api::CreateVersionBody>,
) -> ApiResult<(StatusCode, Json<Created>)> {
    let project = find_project(&state, &project).await?;
    let key = body.key.trim();
    if key.is_empty() {
        return Err(ApiError::BadRequest("key is required".into()));
    }

    let status = body.support_status.unwrap_or(api::SupportStatus::Supported);
    let version = db::create_version(
        &state.db,
        project.id,
        key,
        status.as_str(),
        body.java_min_version,
    )
    .await
    .map_err(|error| match error {
        database::Error::Database(ref inner) if inner.is_unique_violation() => {
            ApiError::Conflict(format!("Version {key} already exists"))
        }
        other => ApiError::Database(other),
    })?;

    Ok((
        StatusCode::CREATED,
        Json(Created {
            message: "Version created".into(),
            version: version.key,
        }),
    ))
}

File: apps/azimuth/src/routes/atlas/versions.ts
import { CreateVersionBodySchema } from '@bx-team/types/schema/atlas';
import { Hono } from 'hono';
import { atlasDb, createVersion, findVersion, listBuilds, listVersions } from '../../database/models/atlas';
import { versionResponse } from '../../models/atlas';
import { badRequest, conflict, internal } from '../../util/error';
import { requireApiSecret } from '../../util/secret';
import { type AtlasEnv, requireProject, requireVersion } from './context';

export const versions = new Hono<AtlasEnv>();

versions.get('/projects/:project/versions', async c => {
  const db = atlasDb(c.env.ATLAS_DB);
  const project = await requireProject(db, c.req.param('project'));

  const projectVersions = await listVersions(db, [project.id]);
  const versionBuilds = await listBuilds(
    db,
    projectVersions.map(version => version.id),
  );

  const buildNumbersByVersion = new Map<number, number[]>();
  for (const build of versionBuilds) {
    buildNumbersByVersion.set(build.versionId, [
      ...(buildNumbersByVersion.get(build.versionId) ?? []),
      build.buildNumber,
    ]);
  }

  return c.json(projectVersions.map(version => versionResponse(version, buildNumbersByVersion.get(version.id) ?? [])));
});

versions.post('/projects/:project/versions/create', requireApiSecret, async c => {
  const db = atlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');

  const parsed = CreateVersionBodySchema.safeParse(await c.req.json().catch(() => ({})));
  if (!parsed.success) throw badRequest(parsed.error.message);
  const body = parsed.data;

  const project = await requireProject(db, projectKey);

  if (await findVersion(db, project.id, body.key)) {
    throw conflict(`Version '${body.key}' already exists for project '${projectKey}'`);
  }

  const version = await createVersion(db, project.id, {
    key: body.key,
    supportStatus: body.supportStatus ?? 'SUPPORTED',
    javaMinVersion: body.javaMinVersion ?? null,
  });
  if (!version) throw internal('Failed to create version');

  return c.json(
    {
      message: 'Version created successfully',
      version: {
        id: version.id,
        project: projectKey,
        key: version.key,
        supportStatus: version.supportStatus,
        javaMinVersion: version.javaMinVersion,
      },
    },
    201,
  );
});

versions.get('/projects/:project/versions/:version', async c => {
  const db = atlasDb(c.env.ATLAS_DB);
  const projectKey = c.req.param('project');

  const project = await requireProject(db, projectKey);
  const version = await requireVersion(db, project.id, projectKey, c.req.param('version'));
  const versionBuilds = await listBuilds(db, [version.id]);

  return c.json(
    versionResponse(
      version,
      versionBuilds.map(build => build.buildNumber),
    ),
  );
});
